### PR TITLE
Slot DAO impl

### PR DESCRIPTION
**Pull request by anastygnome**

### CODE REVIEW

**Comment by Desmondmies** [comment]
in AppLauncher, the line `//HelloApp.main(args)` was for testing purpose and training in javafx, like i said we will delete all javafx test files afterwards so no need to update them 

**HelloApp / HelloController / test-view.fxml / test-view-styles.css**

**Comment by Theo-Hafsaoui** [comment]
oups

File: edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    /**
+     * Getter for one Slot
+     *
+     * @param id numerical long identifier for getting the Slot
+     * @return May return one Slot instance
+     */
+    @Override
+    public Optional<Slot> get ( final long id ) throws SQLException {
+        try (Connection conn = Datasource.getConnection();
+             PreparedStatement stmt = conn.prepareStatement(SELECT_STMT)
```
Comment (Theo-Hafsaoui) [inline edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java:52]: Why doing the prepared statement now, and not when the dao is create ?

File: edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            final Instant fetchedBegin = Instant.ofEpochSecond(rs.getLong(2));
+            final Instant fetchedEnd = Instant.ofEpochSecond(rs.getLong(3));
+            final Interval fetchedInterval = Interval.of(fetchedBegin, fetchedEnd);
+            final Classroom fetchedClassroom = ClassroomDAO.getInstance().get(rs.getLong(4)).orElseThrow(SQLException::new);
+            final String fetchedMemo = rs.getString(5);
+            final Subject fetchedSubject = SubjectDAO.getInstance().get(rs.getLong(6)).orElseThrow(SQLException::new);
+            final Slot.SlotType fetchedType = Slot.SlotType.valueOf(rs.getString(7));
+            final List<Group> fetchedGroup = new ArrayList<>(GroupDAO.getInstance().getGroupOfSlot(id)); //the returned list is unmodifiable, requiring a copy.
+            return Optional.of(Slot.getInstance(fetchedID, fetchedType, fetchedClassroom, fetchedSubject, fetchedGroup, List.of(Professor.of("test","test","test@lhd.org","test")),fetchedInterval, fetchedMemo));
+        } catch (SQLException e) {
```
Comment (Theo-Hafsaoui) [inline edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java:70]: if you want to throw an exception, you should goes to one more level of abstraction.
in other word a user using the dao should not see an ``SQLException``

File: edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    @Override
+    public void delete (final  Slot slot ) {
+        try (Connection conn = Datasource.getConnection();
+             PreparedStatement stmt = conn.prepareStatement(DELETE_STMT)){
+            final long id = slot.getId();
+            if (id<0) { throw new IdException();}
+            stmt.setLong(1,id);
+            if(stmt.executeUpdate() == 0) {
+                throw new IdException("ID" + id + "doesn't exist");
```
Comment (Theo-Hafsaoui) [inline edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java:120]: nice idea, I only thought of IdException as a way to prevent -1 inside the db

File: edtLHD/src/main/java/fr/univtln/lhd/view/edt/EdtGrid.java [hunk 4]
```diff
@@ -16,7 +16,7 @@
 public class EdtGrid extends Grid {
 
     public enum Days { LUNDI, MARDI, MERCREDI, JEUDI, VENDREDI }
-    private List<String> hours;
+    private final List<String> hours;
```
Comment (Theo-Hafsaoui) [inline edtLHD/src/main/java/fr/univtln/lhd/view/edt/EdtGrid.java:19]: not sure if it's a good idea...

**Review by Theo-Hafsaoui — COMMENTED** [review]
In general nice work, see the comment for the point where I have hesitation

File: edtLHD/src/main/java/fr/univtln/lhd/view/edt/EdtGrid.java (hunk 4 above)
Comment (anastygnome) [reply edtLHD/src/main/java/fr/univtln/lhd/view/edt/EdtGrid.java:19]: This is straight from develop. I have not done any changes to this file, @Desmondmies ?

File: edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java (hunk 2 above)
Comment (anastygnome) [reply edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java:70]: Yeah, I wondered whether we should create a Special Exception ("databaseException") or just the SQL error. 
I've left this as is for now.

File: edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java (hunk 1 above)
Comment (Theo-Hafsaoui) [reply edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java:52]: Hmm, correct me if I am wrong,but my comprehension is that if you do a prepared satement
you got one more connnection to the db
```
DAO                 DB
|send prepared stm   |
|------------------->|
|return compile stm  |
|<-------------------|
|                    |
.....after............
|send compile stm    |
|------------------->|
|return result set   |
|<-------------------|

```
I assume that we don't have reason to fear an sql injection, therefore would not it be better to do a simple statement ?

File: edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java (hunk 2 above)
Comment (Theo-Hafsaoui) [reply edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java:70]: k

File: edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java (hunk 1 above)
Comment (anastygnome) [reply edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SlotDAO.java:52]: The setString, setInt and other methods are here to prevent SQL injections. 
### TLDR: here is a statement analogy
> Hey postgres I'm gonna need some id, timestamp, and thing from a table, I'll give you the values  later.
 Postgres can anticipate, as the query planning rarely needs the specific IDs or values.  
### Details 
As the java documentation says : 

The process is as follows : 
1. A connection is opened, in our case, it's just a pop from the set (ConcurrentBag) of HikariCP
2. The String is sent to the database, using a `PREPARE `command, which as per postgresQL documentation [^1]  just sends the String with a placeholder (like the `?` of JDBC)  for the settings, this allows the DBMS to plan the query(ies) as it has the information on rows and columns at hand. This is useful, if part of the parameters in your query rely on values that need time to be computed.
 


4. Once the statement's results are processed, the connection is closed (given back to the pool), when the try block is complete

[^1]:Text above from [PostgreSQL docs](https://www.postgresql.org/docs/15/sql-prepare.html) and [Java docs](https://docs.oracle.com/javase/tutorial/jdbc/basics/prepared.html), respectively
> When the PREPARE statement [statement here means SQL language instruction] is executed, the specified statement is parsed, analyzed, and rewritten. When an EXECUTE [in java ExecuteQuery()] command is subsequently issued, the prepared statement is planned and executed. This division of labor avoids repetitive parse analysis work, while allowing the execution plan to depend on the specific parameter values supplied.


> The main feature of a PreparedStatement object is that, unlike a Statement object, it is given a SQL statement when it is created. The advantage to this is that in most cases, this SQL statement is sent to the DBMS right away, where it is compiled. As a result, the PreparedStatement object contains not just a SQL statement, but a SQL statement that has been precompiled. This means that when the PreparedStatement is executed, the DBMS can just run the PreparedStatement SQL statement without having to compile it first.

**Comment by anastygnome** [comment]
merged in #36
